04.ls/ls.rb

```diff
@@ -2,26 +2,126 @@
 # frozen_string_literal: true
 
 require 'optparse'
+require 'etc'
 
 COLUMNS = 3
 TAB_WIDTH = 8
 
+ENTRY_TYPES = {
+  file: '-',
+  blockSpecial: 'b',
+  characterSpecial: 'c',
+  directory: 'd',
+  link: 'l',
+  fifo: 'p',
+  socket: 's',
+  unknown: 'w'
+}.freeze
+
+FILE_PERMISSIONS = {
+  '0': '---',
+  '1': '--x',
+  '2': '-w-',
+  '3': '-wx',
+  '4': 'r--',
+  '5': 'r-x',
+  '6': 'rw-',
+  '7': 'rwx'
+}.freeze
+
+HALF_A_YEAR_SECONDS = (60 * 60 * 24) * (365.2425 / 2)
+
 def main
-  file_names = search_file_names
+  options = ARGV.getopts('alr')
+  searched_file_names = options['a'] ? Dir['..', '.*', '*'] : Dir['*']
+  sorted_file_names = searched_file_names.sort_by(&:downcase)
+  file_names = options['r'] ? sorted_file_names.reverse : sorted_file_names
+  options['l'] ? display_long_format(file_names) : display_column_format(file_names)
+end
+
+def display_long_format(file_names)
+  long_formats = build_long_format_table(file_names)
+  total_blocks = long_formats.sum { |format| format[:blocks] }
+  puts "total #{total_blocks}"
+
+  widths = caluculate_max_length(long_formats)
+  long_formats.each do |format|
+    rows = [
+      "#{format[:file_mode]} ",
+      format[:hard_links].rjust(widths[:link_width]),
+      "#{format[:owner_name].ljust(widths[:owner_width])} ",
+      "#{format[:group_name].ljust(widths[:group_width])} ",
+      format[:byte_size].rjust(widths[:byte_width]),
+      format[:last_modified_time],
+      format[:path_name]
+    ]
+    puts rows.join(' ')
+  end
+end
+
+def build_long_format_table(file_names)
+  file_names.map do |file_name|
+    status = File.lstat(file_name)
+    {
+      file_mode: "#{ENTRY_TYPES[status.ftype.to_sym]}#{create_permissions(status)}",
+      hard_links: status.nlink.to_s,
+      owner_name: Etc.getpwuid(status.uid).name,
+      group_name: Etc.getgrgid(status.gid).name,
+      byte_size: status.blockdev? || status.chardev? ? format('%#x', status.rdev) : status.size.to_s,
```

Comment:
```suggestion
      byte_size: byte_size,
```

事前に変数に入れておいた方が読みやすいかなーと思います。path_nameも同様です。
https://x.com/jnchito/status/1326468865103073280

Reply:
変数に入れて判定するようにしました。
[サンプルコードでわかる！Ruby 3\.1の主な新機能と変更点 \#Ruby \- Qiita](https://qiita.com/jnchito/items/bcd9b7f59bf4b30ea5b3#%E3%83%8F%E3%83%83%E3%82%B7%E3%83%A5%E3%83%AA%E3%83%86%E3%83%A9%E3%83%AB%E3%81%A8%E3%82%AD%E3%83%BC%E3%83%AF%E3%83%BC%E3%83%89%E5%BC%95%E6%95%B0%E3%81%A7%E5%80%A4%E3%82%92%E7%9C%81%E7%95%A5%E3%81%A7%E3%81%8D%E3%82%8B%E3%82%88%E3%81%86%E3%81%AB%E3%81%AA%E3%81%A3%E3%81%9F)
こちらの記事を読んで、ローカル変数名とハッシュのキー名が同じなので値を省略しました。